master/internal/user/postgres_users.go

```diff
@@ -172,7 +173,12 @@ func DeleteSessionByID(ctx context.Context, sessionID model.SessionID) error {
 // AddUserTx & addAgentUserGroup are helper methods for Add & Update.
 // AddUserTx UPSERT's the existence of a new user.
 func AddUserTx(ctx context.Context, idb bun.IDB, user *model.User) (model.UserID, error) {
-	if _, err := idb.NewInsert().Model(user).Returning("id").Exec(ctx); err != nil {
+	if _, err := idb.NewInsert().Model(user).ExcludeColumn("id").Returning("id").Exec(ctx); err != nil {
+		if pgerr, ok := errors.Cause(err).(*pgconn.PgError); ok {
+			if pgerr.Code == db.CodeUniqueViolation {
+				return 0, db.ErrDuplicateRecord
```

Comment:
question (non-blocking): would we want to wrap this returned error, while also swapping out the error for the `db.ErrorDuplicatedRecord`? Something like this?
```suggestion
				return 0, fmt.Errorf("error inserting user: %s", db.ErrorDuplicatedRecord)
```

Reply:
I don't think we should wrap in this specific instance because we error match in `master/internal/user/service.go`
```
_, err = Add(ctx, &userToAdd, ug)
	switch {
	case err == db.ErrDuplicateRecord:
		return nil, echo.NewHTTPError(http.StatusBadRequest, "user already exists")
	case err != nil:
		return nil, err
	}
```
Pachyderm integration logic searches for this error message "user already exists".

Do I believe this is best practice? Not really, but we are making pre-defined errors for cases like this.

Reply:
Is the Pachyderm integration logic written in Go? 

Wouldn't it make more since to use something like `errors.Is(err, db.ErrDuplicateRecord)`, since it will return true even if the error is wrapped?

Reply:
Yeah, see https://github.com/pachyderm/pachyderm/blob/766571efbb6eba8dd9214e1e2d38f54aca7ea63f/src/server/pps/server/determined.go, in their repo.

I didn't know `errors.Is(err, db.ErrDuplicateRecord)` could do that, handy.

Reply:
I guess I'm confused as far as the best practice question, are you talking about the determined code or are you talking about this line in the Pachyderm integration: 
```
if ... strings.Contains(err.Error(), "user already exists")
```
https://github.com/pachyderm/pachyderm/blob/766571efbb6eba8dd9214e1e2d38f54aca7ea63f/src/server/pps/server/determined.go#L339

I think strings.Contains makes a lot of sense here in response to what we've got going on. If we had a predefined "user already exist" error then that they could import & check against, then `errors.Is()` would be more appropriate. The current code will break if we ever decide to change the exact wording of the error that ../user/service.go is returning. 

----
If you're talking about the determined code, I'm not a big fan of splitting the responsibility of wrapping the errors between two functions within the same "scope" (there's probably a better word for this) since it can muddy the waters when errors need to be wrapped and then lead to  debugging being more difficult since we're missing that info. 

That's being said, I think it can be argued that wrapping `ErrDuplicateRecord` is unnecessary since it's not user facing (beyond the API/CLI) + is an internally defined error that we already have context on when we receive it; however, it's less about this specific error, and more making development & debugging easier to do & think about. 

----
On another note, checking for error equality works only if we never expect the error to be wrapped which isn't always a safe assumption especially when future proofing. I'd encourage using a more flexible approach, like `errors.Is()`, rather than strictly equals unless you explicitly want only that unwrapped error or are trying to improve performance by avoiding the recursive check.

Wrapping `ErrDuplicateRecord` shouldn't an issue for the Pachyderm integration, as long as we are handling the wrapped error properly, since the HTTPError containing `"user already exists"` is only dependent on it being passed in to the constructor & us getting into that case (assuming `==` is replaced with something like `errors.Is()`).